geomesa-memory/src/main/scala/org/locationtech/geomesa/memory/cqengine/utils/CQEngineQueryVisitor.scala

```diff
@@ -0,0 +1,365 @@
+import com.googlecode.cqengine.query.Query
+import com.googlecode.cqengine.{query => cqquery}
+import com.vividsolutions.jts.geom.Geometry
+import org.geotools.filter.LikeToRegexConverter
+import org.geotools.filter.visitor.AbstractFilterVisitor
+import org.locationtech.geomesa.filter._
+import org.locationtech.geomesa.memory.cqengine.query.{GeoToolsFilterQuery, Intersects => CQIntersects}
+import org.opengis.feature.simple.{SimpleFeature, SimpleFeatureType}
+import org.opengis.filter._
+import org.opengis.filter.expression.PropertyName
+import org.opengis.filter.spatial._
+import org.opengis.filter.temporal._
+import org.opengis.temporal.Period
+
+import scala.collection.JavaConversions._
+import scala.language._
+
+class CQEngineQueryVisitor(sft: SimpleFeatureType) extends AbstractFilterVisitor {
+  implicit val lookup = SFTAttributes(sft)
+
+  /* Logical operators */
+
+  /**
+    * And
+    */
+  override def visit(filter: And, data: scala.Any): AnyRef = {
+    val children = filter.getChildren
+
+    val query = children.map { f =>
+      f.accept(this, null) match {
+        case q: Query[SimpleFeature] => q
+        case _ => throw new RuntimeException(s"Can't parse filter: $f.")
+      }
+    }.toList
+    new cqquery.logical.And[SimpleFeature](query)
+  }
+
+  /**
+    * Or
+    */
+  override def visit(filter: Or, data: scala.Any): AnyRef = {
+    val children = filter.getChildren
+
+    val query = children.map { f =>
+      f.accept(this, null) match {
+        case q: Query[SimpleFeature] => q
+        case _ => throw new RuntimeException(s"Can't parse filter: $f.")
+      }
+    }.toList
+    new cqquery.logical.Or[SimpleFeature](query)
+  }
+
+  /**
+    * Not
+    */
+  override def visit(filter: Not, data: scala.Any): AnyRef = {
+    val subfilter = filter.getFilter
+
+    val subquery = subfilter.accept(this, null) match {
+      case q: Query[SimpleFeature] => q
+      case _ => throw new RuntimeException(s"Can't parse filter: $subfilter.")
+    }
+    new cqquery.logical.Not[SimpleFeature](subquery)
+  }
+
+  /* Id, null, nil, exclude, include */
+
+  /**
+    * Id
+    */
+  override def visit(filter: Id, extractData: scala.AnyRef): AnyRef = {
+    val attr = SFTAttributes.fidAttribute
+    val values = filter.getIDs.map(_.toString)
+    new cqquery.simple.In[SimpleFeature, String](attr, true, values)
+  }
+
+  /**
+    * PropertyIsNil:
+    * follows the example of IsNilImpl by using the same implementation as PropertyIsNull
+    */
+  override def visit(filter: PropertyIsNil, extraData: scala.Any): AnyRef = {
+    val prop = getProp(filter)
+    val attr = lookup.lookup[Any](prop.name)
+    new cqquery.logical.Not[SimpleFeature](new cqquery.simple.Has(attr))
+  }
+
+  /**
+    * PropertyIsNull
+    */
+  override def visit(filter: PropertyIsNull, data: scala.Any): AnyRef = {
+    val prop = getProp(filter)
+    val attr = lookup.lookup[Any](prop.name)
+    // TODO: could this be done better?
```

Comment:
We could implement a new CQEngine ``Query`` type, but it would require patching all of the existing CQEngine ``Index``es.